### ISSUE TITLE
[PATCH] Skip main::!->somemethod in B::walksymtable

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#11172]
Migrated from [rt.perl.org#85370](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=85370) (status was 'rejected')

Searchable as RT85370$

**Comment by p5pRT** [comment]
### From @rurban

This is a bug report for perl from rurban@&#8203;x\-ray\.at\,
generated with the help of perlbug 1\.39 running under perl 5\.13\.10\.




---
Calling main&#8203;::\!\->somemethod errors since 5\.13\.10\, skip it\.

See comment in patched source\.


---
 ext/B/B\.pm |    2 \+\+
 1 files changed\, 2 insertions\(\+\)\, 0 deletions\(\-\)

<details><summary>Inline Patch</summary>

```diff
diff --git a/ext/B/B.pm b/ext/B/B.pm
index 8f3c975..b589763 100644
--- a/ext/B/B.pm
+++ b/ext/B/B.pm
@@ -258,6 +258,8 @@ sub walksymtable {
 	    if (svref_2object(\*$sym)->NAME ne "main::" && $sym ne

```
</details>
"<none>::" && &$recurse($sym)) {
&nbsp;                walksymtable\(\\%$fullname\, $method\, $recurse\, $sym\);
&nbsp;         \}
\+    \} elsif \($fullname eq '\*main&#8203;::\!'\) \{
\+            ; \# skip to avoid Errno\, which reloads Config\, which throws
redefinition warnings
&nbsp;     \} else \{
&nbsp;            svref\_2object\(\\\*$fullname\)\->$method\(\);
&nbsp;     \}

The problem was detected in the compiler\, see
http&#8203;://code\.google\.com/p/perl\-compiler/issues/detail?id=63

Unwanted output from the compiler since 5\.13\.10
&nbsp;  Subroutine import redefined at
/usr/lib/perl5/5\.13\.10/i686\-nothreads\-debug\-cygwin/Config\.pm line 38\.
&nbsp;  Subroutine FETCH redefined at
/usr/lib/perl5/5\.13\.10/i686\-nothreads\-debug\-cygwin/Config\.pm line 65\.
&nbsp;  Subroutine TIEHASH redefined at
/usr/lib/perl5/5\.13\.10/i686\-nothreads\-debug\-cygwin/Config\.pm line 72\.
&nbsp;  Subroutine DESTROY redefined at /usr/lib/perl5/5\.13\.10/i686\-nothreads\-
debug\-cygwin/Config\.pm line 76\.

when calling \*main&#8203;::\!\->savecv from B&#8203;::walksymtable
&nbsp;  svref\_2object\(\\\*$fullname\)\->$method\(\)
which steps into Errno \(core error\)\, which does use Config; where %INC has only
Config\_heavy\.pl defined not Config\.pm\, so we get these warnings\.

This is before stepping into our savecv\, so it's useless to filter it there\.
It must be filtered before\, best in B&#8203;::walksymtable



---


---
Flags&#8203;:
&nbsp;    category=library
&nbsp;    severity=low
&nbsp;    module=B


---
This perlbug was built using Perl 5\.10\.1 \- Sat Aug 28 20&#8203;:14&#8203;:06 CEST 2010
It is being executed now by  Perl 5\.13\.10 \- Sat Feb 26 18&#8203;:42&#8203;:59 CET 2011\.

Site configuration information for perl 5\.13\.10&#8203;:

Configured by rurban at Sat Feb 26 18&#8203;:42&#8203;:59 CET 2011\.

Summary of my perl5 \(revision 5 version 13 subversion 10\) configuration&#8203;:
&nbsp;   
&nbsp;  Platform&#8203;:
&nbsp;    osname=cygwin\, osvers=1\.7\.8s\(0\.23153\)\, archname=cygwin\-64int
&nbsp;    uname='cygwin\_nt\-6\.1 atgrzwn503050 1\.7\.8s\(0\.23153\) 20100929 11&#8203;:09&#8203;:59 i686 cygwin '
&nbsp;    config\_args='\-de \-Dlibperl=cygperl5\_13\_10d\-nt\.dll \-Dcc=gcc\-4 \-Dld=g\+\+\-4 \-Accflags=\-march=pentium4 \-Accflags=\-mfpmath=sse \-Accflags=\-mieee\-fp \-Accflags=\-mmmx \-Accflags=\-msse \-Accflags=\-msse2 \-Dmksymlinks \-Uusethreads \-Dmad=y \-Dusedevel \-Dstatic\_ext=Cwd \-DDEBUGGING \-Doptimize=\-g3'
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;    useithreads=undef\, usemultiplicity=undef
&nbsp;    useperlio=define\, d\_sfio=undef\, uselargefiles=define\, usesocks=undef
&nbsp;    use64bitint=define\, use64bitall=undef\, uselongdouble=undef
&nbsp;    usemymalloc=y\, bincompat5005=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='gcc\-4'\, ccflags ='\-DPERL\_USE\_SAFE\_PUTENV \-U\_\_STRICT\_ANSI\_\_ \-march=pentium4 \-mfpmath=sse \-mieee\-fp \-mmmx \-msse \-msse2 \-DDEBUGGING \-fno\-strict\-aliasing \-pipe \-fstack\-protector \-I/usr/local/include'\,
&nbsp;    optimize='\-g3'\,
&nbsp;    cppflags='\-DPERL\_USE\_SAFE\_PUTENV \-U\_\_STRICT\_ANSI\_\_ \-march=pentium4 \-mfpmath=sse \-mieee\-fp \-mmmx \-msse \-msse2 \-DDEBUGGING \-fno\-strict\-aliasing \-pipe \-fstack\-protector \-I/usr/local/include'
&nbsp;    ccversion=''\, gccversion='4\.3\.4 20090804 \(release\) 1'\, gccosandvers=''
&nbsp;    intsize=4\, longsize=4\, ptrsize=4\, doublesize=8\, byteorder=12345678
&nbsp;    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=12
&nbsp;    ivtype='long long'\, ivsize=8\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\, lseeksize=8
&nbsp;    alignbytes=8\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='g\+\+\-4'\, ldflags =' \-Wl\,\-\-enable\-auto\-import \-Wl\,\-\-export\-all\-symbols \-Wl\,\-\-enable\-auto\-image\-base \-fstack\-protector \-L/usr/local/lib'
&nbsp;    libpth=/usr/local/lib /usr/lib /lib
&nbsp;    libs=\-lgdbm \-ldb \-ldl \-lcrypt \-lgdbm\_compat
&nbsp;    perllibs=\-ldl \-lcrypt
&nbsp;    libc=/usr/lib/libc\.a\, so=dll\, useshrplib=true\, libperl=cygperl5\_13\_10d\-nt\.dll
&nbsp;    gnulibc\_version=''
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_dlopen\.xs\, dlext=dll\, d\_dlsymun=undef\, ccdlflags=' '
&nbsp;    cccdlflags=' '\, lddlflags=' \-\-shared  \-Wl\,\-\-enable\-auto\-import \-Wl\,\-\-export\-all\-symbols \-Wl\,\-\-enable\-auto\-image\-base \-L/usr/local/lib \-fstack\-protector'

Locally applied patches&#8203;:
&nbsp;    CYG11 no\-bs
&nbsp;    CYG12 no archlib in otherlibdirs
&nbsp;    CYG14 Dynaloader
&nbsp;    CYG15 static\-Win32CORE
&nbsp;    CYG17 utf8\-paths
&nbsp;    CYG21 LibList\-Kid\.patch
&nbsp;    CYG22 cygwin\-1\.7 hints
&nbsp;    CYG23 544\-stat
&nbsp;    CYG24 build man pages
&nbsp;    CYG25 rebase\_privlib
&nbsp;    Module\-Build\-0\.36\_13
&nbsp;    Bug\#55162 CYG18 File&#8203;::Spec&#8203;::case\_tolerant performance
&nbsp;    disable ExtUtils&#8203;::MakeMaker&#8203;::Coverage in Sys\-Syslog



---
@&#8203;INC for perl 5\.13\.10&#8203;:
&nbsp;    /usr/lib/perl5/site\_perl/5\.13\.10/i686\-nothreads\-debug\-cygwin
&nbsp;    /usr/lib/perl5/site\_perl/5\.13\.10
&nbsp;    /usr/lib/perl5/vendor\_perl/5\.13\.10/i686\-nothreads\-debug\-cygwin
&nbsp;    /usr/lib/perl5/vendor\_perl/5\.13\.10
&nbsp;    /usr/lib/perl5/5\.13\.10/i686\-nothreads\-debug\-cygwin
&nbsp;    /usr/lib/perl5/5\.13\.10
&nbsp;    /usr/lib/perl5/site\_perl
&nbsp;    /usr/lib/perl5/vendor\_perl
&nbsp;    /usr/lib/perl5/site\_perl/5\.13\.9
&nbsp;    /usr/lib/perl5/site\_perl/5\.13\.8
&nbsp;    /usr/lib/perl5/site\_perl/5\.13\.7
&nbsp;    /usr/lib/perl5/site\_perl/5\.13\.6
&nbsp;    /usr/lib/perl5/site\_perl/5\.13\.5
&nbsp;    /usr/lib/perl5/site\_perl/5\.12
&nbsp;    /usr/lib/perl5/site\_perl/5\.10
&nbsp;    /usr/lib/perl5/vendor\_perl/5\.10
&nbsp;    /usr/lib/perl5/site\_perl/5\.8
&nbsp;    \.



---
Environment for perl 5\.13\.10&#8203;:
&nbsp;    HOME=/home/urbanr
&nbsp;    LANG=C\.UTF\-8
&nbsp;    LANGUAGE \(unset\)
&nbsp;    LD\_LIBRARY\_PATH \(unset\)
&nbsp;    LOGDIR \(unset\)
&nbsp;    PATH=/home/urbanr/bin&#8203;:\~/bin&#8203;:/usr/local/bin&#8203;:/usr/bin&#8203;:/usr/sbin&#8203;:/cygdrive/c/strawberry/c/bin&#8203;:/cygdrive/c/WINDOWS/System32&#8203;:/cygdrive/c/WINDOWS&#8203;:/cygdrive/c/WINDOWS/System32/WBEM&#8203;:/cygdrive/c/PROGRA~1/ATT/Graphviz/bin&#8203;:/cygdrive/c/PROGRA~1/ATT/Graphviz/bin/tools&#8203;:/usr/lib/lapack&#8203;:/cygdrive/c/oracle/ora112/bin
&nbsp;    PERL\_BADLANG \(unset\)
&nbsp;    SHELL \(unset\)

**Comment by p5pRT** [comment]
### From @rurban

patch attached
\-\- 
Reini Urban

**Comment by p5pRT** [comment]
### From @rurban

<details><summary>0001-Calling-main-somemethod-errors-since-5.13.10-skip-it.patch</summary>

```diff
From b50bd97420fdd0e19b75637a5c68b4653373407f Mon Sep 17 00:00:00 2001
From: Reini Urban <rurban@x-ray.at>
Date: Thu, 3 Mar 2011 17:01:33 +0100
Subject: [PATCH] Calling main::!->somemethod errors since 5.13.10, skip it.

See comment in patched source.
---
 ext/B/B.pm |    2 ++
 1 files changed, 2 insertions(+), 0 deletions(-)

diff --git a/ext/B/B.pm b/ext/B/B.pm
index 8f3c975..b589763 100644
--- a/ext/B/B.pm
+++ b/ext/B/B.pm
@@ -258,6 +258,8 @@ sub walksymtable {
 	    if (svref_2object(\*$sym)->NAME ne "main::" && $sym ne "<none>::" && &$recurse($sym)) {
                walksymtable(\%$fullname, $method, $recurse, $sym);
 	    }
+	} elsif ($fullname eq '*main::!') {
+            ; # skip to avoid Errno, which reloads Config, which throws redefinition warnings
 	} else {
            svref_2object(\*$fullname)->$method();
 	}
-- 
1.7.4


```
</details>

**Comment by p5pRT** [comment]
@rurban - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @cpansprout

On Thu Mar 03 08&#8203;:06&#8203;:53 2011\, rurban wrote&#8203;:
> This is a bug report for perl from rurban@&#8203;x\-ray\.at\,
> generated with the help of perlbug 1\.39 running under perl 5\.13\.10\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> Calling main&#8203;::\!\->somemethod errors since 5\.13\.10\, skip it\.
> 
> See comment in patched source\.
> \-\-\-
>  ext/B/B\.pm |    2 \+\+
>  1 files changed\, 2 insertions\(\+\)\, 0 deletions\(\-\)
> 
> diff \-\-git a/ext/B/B\.pm b/ext/B/B\.pm
> index 8f3c975\.\.b589763 100644
> \-\-\- a/ext/B/B\.pm
> \+\+\+ b/ext/B/B\.pm
> @&#8203;@&#8203; \-258\,6 \+258\,8 @&#8203;@&#8203; sub walksymtable \{
>          if \(svref\_2object\(\\\*$sym\)\->NAME ne "main&#8203;::" && $sym ne
> "\<none>&#8203;::" && &$recurse\($sym\)\) \{
>                 walksymtable\(\\%$fullname\, $method\, $recurse\, $sym\);
>          \}
> \+    \} elsif \($fullname eq '\*main&#8203;::\!'\) \{
> \+            ; \# skip to avoid Errno\, which reloads Config\, which
>    throws
> redefinition warnings
>      \} else \{
>             svref\_2object\(\\\*$fullname\)\->$method\(\);
>      \}

Would it not be better to skip it only if Errno\.pm is not already
loaded? Could you also provide a test case?

**Comment by p5pRT** [comment]
### From @rurban

2011/3/6 Father Chrysostomos via RT \<perlbug\-followup@&#8203;perl\.org>&#8203;:
> On Thu Mar 03 08&#8203;:06&#8203;:53 2011\, rurban wrote&#8203;:
>> This is a bug report for perl from rurban@&#8203;x\-ray\.at\,
>> generated with the help of perlbug 1\.39 running under perl 5\.13\.10\.
>>
>>
>> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
>> Calling main&#8203;::\!\->somemethod errors since 5\.13\.10\, skip it\.
>>
>> See comment in patched source\.
>> \-\-\-
>>  ext/B/B\.pm |    2 \+\+
>>  1 files changed\, 2 insertions\(\+\)\, 0 deletions\(\-\)
>>
>> diff \-\-git a/ext/B/B\.pm b/ext/B/B\.pm
>> index 8f3c975\.\.b589763 100644
>> \-\-\- a/ext/B/B\.pm
>> \+\+\+ b/ext/B/B\.pm
>> @&#8203;@&#8203; \-258\,6 \+258\,8 @&#8203;@&#8203; sub walksymtable \{
>>           if \(svref\_2object\(\\\*$sym\)\->NAME ne "main&#8203;::" && $sym ne
>> "\<none>&#8203;::" && &$recurse\($sym\)\) \{
>>                 walksymtable\(\\%$fullname\, $method\, $recurse\, $sym\);
>>           \}
>> \+     \} elsif \($fullname eq '\*main&#8203;::\!'\) \{
>> \+            ; \# skip to avoid Errno\, which reloads Config\, which
>>    throws
>> redefinition warnings
>>       \} else \{
>>             svref\_2object\(\\\*$fullname\)\->$method\(\);
>>       \}
>
> Would it not be better to skip it only if Errno\.pm is not already
> loaded? Could you also provide a test case?

You are right\. While working on the test case I found a better fix within B&#8203;::C
so core walksymtable can happily step into $\! hashes without any noise\.

Preloading Errno is the trick\.
Please reject
\-\- 
Reini Urban

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'rejected'
